apt/private/package_resolution.bzl

```diff
@@ -1,165 +0,0 @@
-"package resolution"
-
-load(":version.bzl", version_lib = "version")
```

Comment:
i don't see how it benefits us to have this file merged into package_index. i'd rather have parsing and resolution done separately.  unless you have a really good reason, i'd say we revert this change.

Reply:
I thought the explanation in the commit was a good reason! 😅   Let me copy-paste it here and further explain my train of thought, in case it's helpful.

Previously, in `resolve.bzl` (and IMO the name is already hinting something 😄) we had:
```starlark
    pkgindex = package_index.new(rctx, sources = sources, archs = manifest["archs"])
    pkgresolution = package_resolution.new(index = pkgindex)
    
    constraint = package_resolution.parse_depends(dep_constraint).pop()
```

And now you have:
```starlark
    pkgindex = package_index.new(rctx, sources = sources, archs = manifest["archs"])
    (...)
    constraint = package_index.parse_depends(dep_constraint).pop()
```

`package_resolution` was used only in `resolve.bzl` and only after initializing the `pkgindex`. All of the stuff in `pkgindex` is internal state used to resolve the packages. That's why I think it should all be part of `package_index` which encompasses all of the Debian index functionality.

I think this is nicely shown in the new `_new()`:
```starlark
def _new(rctx, sources, archs):
    index = _index(rctx, sources, archs)

    return struct(
        resolve_all = lambda **kwargs: _resolve_all(index, **kwargs),
        resolve_package = lambda **kwargs: _resolve_package(index, **kwargs),
    )
```

Here, the `index` `struct` is internal and the only thing exposed is "the API", `resolve_all` and `resolve_package`. Which is what we use "outside", because we fetch and parse the index from the manifest and then, we use that state to resolve the packages and populate the lock.

All in all, it's (1) code layout, one file or two files and (2) code architecture / dependency. Before, I think `index` was leaking, because to me it's clearly internal state. IMHO, If we go back to two files, I would still prefer to have the old `package_index` not used in `resolve` and have `package_resolution` use it and only expose the `resolve_*` "API". But then, I think having it all in one file is easier, especially when it's only ~300 lines of code.

Finally, I think that the testing also points at the "internal state" part that I was mentioning. Right now, you pass a mock manifest and then the internals "take care of the index". If you split it, you have to pass the manifest to test `package_index` and then, you would have to either maintain a mock for the `index` `struct` or... you would have to generate it from the mock manifest by calling `package_index.new()` :)

Anyway, let me know what you think!

Reply:
In fact, I've just re-checked and I think my point can be made much clearer just looking at the current code! 😅 
```starlark
def _create_resolution(index):
    state = struct(index = index)
    return struct(
        resolve_all = lambda **kwargs: _resolve_all(state, **kwargs),
        resolve_package = lambda **kwargs: _resolve_package(state, **kwargs),
    )
```

It's called state :) `package_resolution` is literally receiving the state from outside instead of receiving `sources` and `archs` and generating it internally, like I do :)

Reply:
i still don't understand how merging these two together help us as the original thought was that `package_index` is a big state class that holds information in well-structured format for fast access and resolution is a specific algorithm that uses the package_index to perform the resolution using the information from package index.

Currently, you can do this. 

```
index = package_index.new(...)

resolution_a = package_resolution.new(index)

resolution_b = package_resolution.new(index)
```

but once we merge them into one, we automatically lose the ability, and boundaries between the two. 

> It's called state 

We'll it's a big confusing name to use in the case of `package_index`, it's just that way for consistency, it easily could have been index being passed around. 

I am open to renaming some things here and there but i don't see how merging them into one will help us.

Reply:
In fact, i hope accidental direct commit to main makes things clearer; https://github.com/GoogleContainerTools/rules_distroless/commit/a1b5fe36fa7e1b6b6ffc97e90ccf00f6778fdb72

Reply:
> In fact, i hope accidental direct commit to main makes things clearer; [a1b5fe3](https://github.com/GoogleContainerTools/rules_distroless/commit/a1b5fe36fa7e1b6b6ffc97e90ccf00f6778fdb72)

I'm already having quite a big of a rebase with #59... 😭 😭 😭  I'll see if it makes things clearer... 

To my point earlier, I think the (IMHO) mixed mess of:

https://github.com/GoogleContainerTools/rules_distroless/blob/a1b5fe36fa7e1b6b6ffc97e90ccf00f6778fdb72/apt/tests/resolution_test.bzl#L108-L125

and

https://github.com/GoogleContainerTools/rules_distroless/blob/a1b5fe36fa7e1b6b6ffc97e90ccf00f6778fdb72/apt/private/apt_deb_repository.bzl#L142-L160

show exactly what I "predicted" in my comment:

> Finally, I think that the testing also points at the "internal state" part that I was mentioning. Right now, you pass a mock manifest and then the internals "take care of the index". If you split it, you have to pass the manifest to test package_index and then, you would have to either maintain a mock for the index struct or... you would have to generate it from the mock manifest by calling package_index.new() :)

Having to expose internals of (using the old names) `package_index`, mocking them in "prod code" (having the mock in `package_index.bzl` to be able to call `_methods`, etc... to me, this is a pretty big smell.

> i still don't understand how merging these two together help us as the original thought was that package_index is a big state class that holds information in well-structured format for fast access and resolution is a specific algorithm that uses the package_index to perform the resolution using the information from package index.

Precisely! it's an implementation detail, so it's internal and as such, you test it via `resolution`.

Anyway, the rebase with the changes you made in $59 is huge... I'll have to rewrite the PR anyway. So, I'll leave the merging of this stuff for the end or just leave it out completely.

Reply:
Finally, e.g. from a very quick look, `version_constraints` also has quite a bit of stuff that should be in `version.bzl` (like I actually did, the `relop` stuff). Plus renaming tests from #59 e.g. `resolution_test.bzl` IMHO should be named like the file it tests, and similarly, the `version_constraint` testing should go to its own test file, which I have also done.

I'll try to get another version of this PR done for review over the week, but meanwhile could we coordinate a bit to avoid breaking the PR? I've been holding for months on this PRs and having such breaks makes me have to re-do not just this PR but the stack... 🙏  

Thanks!

Reply:
@thesayyn OK, reading the renaming you did in a1b5fe3 plus

> the original thought was that package_index is a big state class that holds information in well-structured format for fast access and resolution is a specific algorithm that uses the package_index to perform the resolution using the information from package index.

I'll clean up around this and leave them separate, and as mentioned, I'll rename test files, etc, for tests to match the files they test, etc.